ApiData.ts

```diff
@@ -0,0 +1,269 @@
+			});
+	}
+
+	/** Use the fetchData() func to make an HTTP request to the API endpoint and process the data */
+	fetchColumns(): Promise<void> {
+		return this.fetchStrData('http://localhost:2050/columns')
+			.then((response:string) => {
+				const res = JSON.parse(response);
+				this.columnNames = res.map((columnName: string) => ({ name: columnName }));
+				// Initialize the 'data' property as an empty array of GridData objects
+				this.data = new Array<GridData>(this.columnNames.length);
+			})
+			.catch(error => {
+				console.error('Failed to fetch columns:' + error);
+				throw ('Failed to fetch columns:' + error);
+			});
+	}
+
+	/** Get records from API for fetch and search functionality */
+	fetchAndProcessRecords(from: number, to: number): Promise<GridData[]> {
+		$('#spinner').show();
+		$('#grid').hide();
+
+		return this.fetchStrData(`http://localhost:2050/records?from=${from}&to=${to}`)
+			.then((response:string) => {
+				const res = JSON.parse(response);
+				const processedData = res.map((record: string) => {
+					const obj: GridData = {};
+					for (let j = 0; j < this.columnNames.length && j < record.length; j++) {
+						obj[this.columnNames[j].name] = record[j];
+					}
+					return obj;
+				});
+				$('#spinner').hide();
+				$('#grid').show();
+				return processedData;
+			})
+			.catch(error => {
+				console.error('Failed to fetch records: ', error);
+				throw new Error('Failed to fetch records: ' + error);
+			});
+	}
+
+	/** Fetches records using fetchAndProcessRecords(), processes them, displays them, and updates page information. */
+	fetchAndDisplayRecords(): Promise<void> {
+		this.maxRange = this.totalItems - 1;
+		let from = this.firstVal;
+		let to = Math.min(from + this.pageSize, this.maxRange);
+
+		if (to >= this.maxRange) {
+			const lastPage = Math.floor(this.maxRange / this.pageSize) + 1;
+			this.currentPage = lastPage;
+			from = this.maxRange - this.pageSize;
+			to = this.maxRange;
+		}
+
+		return this.fetchAndProcessRecords(from, to)
+			.then(processedData => {
+				this.data = processedData;
+				this.displayRecords();
+			})
+			.catch(error => {
+				console.error('Failed to fetch records:', error);
+				alert('Error occured while fetching records!');
+			});
+	}
+
+	/** search through records using fromID */
+	searchRecords(searchValue: number): Promise<void> {
+		// Maximum allowed Value
+		const maxRange = this.maxRange;
+
+		if (searchValue >= 0 && searchValue <= maxRange) {
+			let from = searchValue;
+			const pageSize = this.pageSize;
+
+			if (from + pageSize > maxRange) {
+				from = Math.max(0, maxRange - pageSize);
+			}
+
+			return this.fetchAndProcessRecords(from, from + pageSize)
+				.then(processedData => {
+					this.data = processedData;
+					this.currentPage = Math.ceil(from / this.pageSize) + 1;
+					// Set firstVal to searched value
+					this.firstVal = from;
+					// Calculate lastVal based on pageSize
+					this.lastVal = from + this.pageSize - 1;
```

Comment:
With the above comment being said. This could then be executed before the time and life would be good, right?